feat: Status-Transition Flows (#2037)
cap-js/docs: guides/flows.md

[[toc]]

## Enabling Flows

Status-transition flows are part of the CAP Node.js core (`@sap/cds`).
Hence, there are no steps required to enable the feature.

To get started with the flow feature in CAP Java, simply add the [cds-feature-flow](https://central.sonatype.com/artifact/com.sap.cds/cds-feature-flow) dependency to your `srv/pom.xml` file:

```xml
<dependency>
  <groupId>com.sap.cds</groupId>
  <artifactId>cds-feature-flow</artifactId>
  <scope>runtime</scope>
</dependency>
```

## Modeling Flows

The following shows the simplest way to model a flow.
The flow is taken from [@capire/xtravels](https://github.com/capire/xtravels).
The annotations in the service model are enough to model the flow so it can be used.

![](./assets/flows/xtravels-flow-simple.svg)

```cds
service TravelService {

  entity Travels as projection on db.Travels
  // Define actions
  actions {
    action acceptTravel();
    action rejectTravel();
    action deductDiscount( percent: Percentage not null ) returns Travels;
  };

  // Define flow (+ status check for "deductDiscount")
  annotate Travels with @flow.status: Status actions {

>     acceptTravel    @from: #Open  @to: #Accepted;

Author: it's in the collapsed `db/schema.cds` above. maybe it wasn't pushed yet...

It ensures transitions are explicitly modeled, validated, and executed in a controlled and reliable way.
In case of more complex requirements, you can extend the flows with custom event handlers.

[[toc]]

## Enabling Flows

Status-transition flows are part of the CAP Node.js core (`@sap/cds`).
Hence, there are no steps required to enable the feature.

To get started with the flow feature in CAP Java, simply add the [cds-feature-flow](https://central.sonatype.com/artifact/com.sap.cds/cds-feature-flow) dependency to your `srv/pom.xml` file:

```xml
<dependency>
  <groupId>com.sap.cds</groupId>
  <artifactId>cds-feature-flow</artifactId>
  <scope>runtime</scope>
</dependency>
```

## Modeling Flows

The following shows the simplest way to model a flow.
The example is taken from [@capire/xtravels](https://github.com/capire/xtravels).
The annotations in the service model are enough to model the flow so it can be used.

![](./assets/flows/xtravels-flow-simple.svg)

The following is an extract of the relevant parts of the domain model:

::: details `db/schema.cds`
```cds [db/schema.cds]
// db/schema.cds
namespace sap.capire.travels;

entity Travels : managed {
  // [...]

>   Status       : Association to TravelStatus default 'O';

Author: you mean `status : String enum {...};`? i get your point, but i also see the value in just copying from xtravels to stay as close to the sample as possible...

btw, `@flow.status` is not needed here

#### `@to`

- Defines the desired target state of the entity after executing the action
- Changes the state of the entity to the value defined in this annotation after executing the action
- Must be a single value from the status enum

Both annotations are optional, but at least one is required to mark an action as a flow action. You can use either one or both, depending on the desired behavior. If you use both, you don't need to implement any handlers for the actions, as all generic handlers are registered automatically.

### Generic Handlers

No handler implementations are required for the actions, as the generic handlers are registered automatically.

#### `before`

From the `@from` annotation, a handler is registered that performs the check described above.
The default `before` handler will validate that your entity instance is in a valid entry state, meaning the current state of the entity must match one of the states specified in the `@from` annotation.
If the entity is not in a valid state, the request fails with a suitable error.
A validation error will return a `409` HTTP Status Code.

#### `on`

An empty handler is registered if none is provided.

> That is, The default `On` handler will complete the action for void return types.

Author: outdated

This ensures the request passes through the generic handler stack.

#### `after`

From the `@to` annotation, a handler is registered that changes the state of the entity to the desired target state.
The default `after` handler will process the transition by automatically updating the status element of the entity to the target state defined in the `@to` annotation.
For example, if the current state is `Open` and the target state is `Accepted`, the `after` handler will update the status to `Accepted` after the action is executed.
This ensures that the state transition is consistently applied without requiring custom logic.

::: tip Generic handlers are not executed for drafts.
For example, calling `acceptTravel()` on a `Travels` entity that is currently being _edited_ has no effect.
:::

> ### `$flow.previous`

Reviewer: ```suggestion
### Reverting to Previous State
```

  entity Travels as projection on db.Travels
  actions {
    action reviewTravel();
    action reopenTravel();
    action blockTravel();
    action unblockTravel();
    action acceptTravel();
    action rejectTravel();
    action deductDiscount( percent: Percentage not null ) returns Travels;
  };

  // Define flow incl. "unblockTravel" that transitions to the previous state
  annotate Travels with @flow.status: Status actions {
    reviewTravel    @from: #Open               @to: #InReview;       // [!code highlight]
    reopenTravel    @from: #InReview           @to: #Open;           // [!code highlight]
    blockTravel     @from: [#Open, #InReview]  @to: #Blocked;        // [!code highlight]
    unblockTravel   @from: #Blocked            @to: $flow.previous;  // [!code highlight]
    acceptTravel    @from: #InReview           @to: #Accepted;
    rejectTravel    @from: #InReview           @to: #Canceled;
    deductDiscount  @from: #Open;
  };

}
```

To transition to the previous state, each entity with a flow that includes at least one transition to `$flow.previous` is automatically augmented with the necessary data structure to record the transitions.

> Specifically, the entities are appended with the aspect `sap.common.FlowHistory`.

Reviewer: ```suggestion
Specifically, the entities are extended with the aspect `sap.common.FlowHistory`, which includes `transitions_` composition that captures the history of state transitions.
```

::: tip Transitions are excluded from projections.
The `transitions_` composition that is automatically appended to the base entity, is also automatically excluded from all projections.
:::

## Extending Flows

The flow annotations can be used for all basic flows that should be modelled, when there is the need for more complex cases, they can be implemented with custom event handlers.

Let's look at some use-cases for extending flows with custom event handlers:
- If the entry state validation depends on additional conditions, you could implement the logic in a custom `before` handler. 
- If the action's return type is anything other than `void`, you must implement a custom `on` handler.

> - If you have multiple target states depending on certain conditions, you must implement a custom `on` handler, and you must **not** use the `@flow.to` annotation.

Reviewer: ```suggestion
- If you have multiple target states depending on certain conditions, you must implement a custom `on` handler, and you must **not** use the `@to` annotation.
```

> - If you want to run logic like contacting an external system on a state transition, consider implementing a custom `on` handler for the respective action.

Reviewer: ```suggestion
```

(You can always do this in a custom handler)
